### PR TITLE
Cleanup and port to JS

### PR DESCRIPTION
**Pull request by jkrems**
The code base is really hard to understand right now. Let's fix that.

### CODE REVIEW

**Comment by khoomeister** [comment]
FTW :fireworks: can't wait to review this one!

**Comment by jkrems** [comment]
Alright! This should be ready for review now.

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,5 @@
 language: node_js
 node_js:
   - "0.10"
-  - "0.12"
```
Comment (khoomeister) [inline .travis.yml]: No 0.12?

**Comment by khoomeister** [comment]
:shipit:

**Pull request merged**